weather_dl/download_pipeline/partition_test.py

```diff
@@ -0,0 +1,262 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import itertools
+import json
+import tempfile
+import typing as t
+import unittest
+from unittest.mock import MagicMock
+
+import apache_beam as beam
+from xarray_beam._src.test_util import EagerPipeline
+
+from .manifest import MockManifest, Location, LocalManifest
+from .parsers import get_subsections
+from .partition import skip_partition, PartitionConfig
+from .stores import InMemoryStore, Store
+
+
+class OddFilesDoNotExistStore(InMemoryStore):
+    def __init__(self):
+        super().__init__()
+        self.count = 0
+
+    def exists(self, filename: str) -> bool:
+        ret = self.count % 2 == 0
+        self.count += 1
+        return ret
+
+
+class PreparePartitionTest(unittest.TestCase):
+
+    def setUp(self) -> None:
+        self.dummy_manifest = MockManifest(Location('mock://dummy'))
+
+    def create_partition_configs(self, config, store: t.Optional[Store] = None) -> t.List[t.Dict]:
+        subsections = get_subsections(config)
+        params_cycle = itertools.cycle(subsections)
+
+        return (EagerPipeline()
+                | beam.Create([config])
+                | PartitionConfig(store, params_cycle, self.dummy_manifest))
+
+    def test_partition_single_key(self):
+        config = {
+            'parameters': {
+                'partition_keys': ['year'],
+                'target_path': 'download-{}.nc',
+            },
+            'selection': {
+                'features': ['pressure', 'temperature', 'wind_speed_U', 'wind_speed_V'],
+                'month': [str(i) for i in range(1, 13)],
+                'year': [str(i) for i in range(2015, 2021)]
+            }
+        }
+
+        actual = self.create_partition_configs(config)
+
+        self.assertListEqual([d['selection'] for d in actual], [
+            {**config['selection'], **{'year': [str(i)]}}
```

Comment:
I feel with statements like this, it might soon be that it is time to make the config a real class instead of a dict of dicts of lists of dicts or whatever the actual type is now.

Reply:
Created an issue to track: #118.